src/concrete_solve.jl

```diff
@@ -305,6 +323,67 @@ function DiffEqBase._concrete_solve_adjoint(prob, alg,
                     end
                 end
             end
+            if !(ArrayInterfaceCore.ismutable(u0))
+                return _out
+            end
+        end
+
+        function df(u, p, t, i;outtype=nothing)
```

Comment:
to make this work, you probably need to change the dgdu accmulation code in adjoint_common inside of the `affect!` definition

Reply:
Oh, so its inferred as a redefinition of the first df, which is causing the tests using the first one to fail